### PR TITLE
Update the State RFC to match the current database format

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

This PR brings Zebra's State RFC up-to-date with the current database format. (Including upcoming Sprout changes.)

This is unexpected client planning work in sprint 24.

## Review

Anyone can review this low-priority PR.

### Reviewer Checklist

  - [x] Design makes sense
  - [x] Design matches implementation

## Follow Up Work

- #3137

### CODE REVIEW

File: book/src/dev/rfcs/0005-state-updates.md [hunk 1]
```diff
@@ -600,29 +600,46 @@ order on byte strings is the numeric ordering).
...
+| `sprout_nullifiers`            | `sprout::Nullifier`    | `()`                                 | Never   |
+| `sprout_anchors`               | `sprout::tree::Root`   | `()`                                 | Never   |
+| `sprout_note_commitment_tree`  | `block::Height`        | `sprout::tree::NoteCommitmentTree`   | Delete  |
+| `sapling_nullifiers`           | `sapling::Nullifier`   | `()`                                 | Never   |
+| `sapling_anchors`              | `sapling::tree::Root`  | `()`                                 | Never   |
+| `sapling_note_commitment_tree` | `block::Height`        | `sapling::tree::NoteCommitmentTree`  | Delete  |
+| `orchard_nullifiers`           | `orchard::Nullifier`   | `()`                                 | Never   |
+| `orchard_anchors`              | `orchard::tree::Root`  | `()`                                 | Never   |
+| `orchard_note_commitment_tree` | `block::Height`        | `orchard::tree::NoteCommitmentTree`  | Delete  |
+| `history_tree`                 | `block::Height`        | `zcash_history::Entry`               | Delete  |
```
Comment (conradoplg) [inline book/src/dev/rfcs/0005-state-updates.md]:
```suggestion
| `history_tree`                 | `block::Height`        | `NonEmptyHistoryTree`               | Delete  |
```

File: book/src/dev/rfcs/0005-state-updates.md [hunk 2]
```diff
@@ -600,29 +600,46 @@ order on byte strings is the numeric ordering).
...
+
+We use big-endian encoding for keys, to allow database index prefix searches.
+
+Amounts:
+- `Amount`: 64 bits, little-endian, signed
+- `ValueBalance`: `[Amount; 4]`
+
+Derived Formats:
+- `*::NoteCommitmentTree`: `bincode` using `serde`
+- `zcash_history::Entry`: `bincode` using `serde`, using `zcash_history`'s `serde` implementation
```
Comment (conradoplg) [inline book/src/dev/rfcs/0005-state-updates.md]:
```suggestion
- `NonEmptyHistoryTree`: `bincode` using `serde`, using `zcash_history`'s `serde` implementation
```

**Review by conradoplg — CHANGES_REQUESTED** [review]
Looks good, there are a couple of suggestions for the history tree which is the encoding of a `NonEmptyHistoryTree` struct

File: book/src/dev/rfcs/0005-state-updates.md [hunk 3]
```diff
@@ -600,29 +600,46 @@ order on byte strings is the numeric ordering).
...
+| `sprout_nullifiers`            | `sprout::Nullifier`    | `()`                                 | Never   |
+| `sprout_anchors`               | `sprout::tree::Root`   | `()`                                 | Never   |
+| `sprout_note_commitment_tree`  | `block::Height`        | `sprout::tree::NoteCommitmentTree`   | Delete  |
+| `sapling_nullifiers`           | `sapling::Nullifier`   | `()`                                 | Never   |
+| `sapling_anchors`              | `sapling::tree::Root`  | `()`                                 | Never   |
+| `sapling_note_commitment_tree` | `block::Height`        | `sapling::tree::NoteCommitmentTree`  | Delete  |
+| `orchard_nullifiers`           | `orchard::Nullifier`   | `()`                                 | Never   |
+| `orchard_anchors`              | `orchard::tree::Root`  | `()`                                 | Never   |
+| `orchard_note_commitment_tree` | `block::Height`        | `orchard::tree::NoteCommitmentTree`  | Delete  |
+| `history_tree`                 | `block::Height`        | `NonEmptyHistoryTree`               | Delete  |
```
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]:
```suggestion
| `orchard_note_commitment_tree` | `block::Height`        | `orchard::tree::NoteCommitmentTree`  | Delete  |
| `history_tree`                 | `block::Height`        | `NonEmptyHistoryTree`                | Delete  |
```

**Review by conradoplg — APPROVED** [review]

**Comment by conradoplg** [comment]
I'll leave syncing up to you since there seems to be other stuff being merged right now

**Pull request merged**